dandiapi/zarr/checksums.py

```diff
@@ -143,38 +147,44 @@ class ZarrChecksumModificationQueue:
     """
 
     def __init__(self) -> None:
-        self._queue: dict[Path, ZarrChecksumModification] = {}
+        self._heap: list[tuple[int, ZarrChecksumModification]] = []
+        self._path_map: dict[Path, ZarrChecksumModification] = {}
+
+    def _add_path(self, key: Path):
+        modification = ZarrChecksumModification(path=key)
+
+        # Add link to modification
+        self._path_map[key] = modification
+
+        # Add modification to heap with length (negated to representa max heap)
+        length = len(key.parents)
+        heapq.heappush(self._heap, (-1 * length, modification))
+
+    def _get_path(self, key: Path):
+        if key not in self._path_map:
+            self._add_path(key)
+
+        return self._path_map[key]
 
     def queue_file_update(self, key: Path, checksum: ZarrChecksum):
-        if key not in self._queue:
-            self._queue[key] = ZarrChecksumModification(path=key)
-        self._queue[key].files_to_update.append(checksum)
+        self._get_path(key).files_to_update.append(checksum)
 
     def queue_directory_update(self, key: Path, checksum: ZarrChecksum):
-        if key not in self._queue:
-            self._queue[key] = ZarrChecksumModification(path=key)
-        self._queue[key].directories_to_update.append(checksum)
+        self._get_path(key).directories_to_update.append(checksum)
 
     def queue_removal(self, key: Path, path: Path | str):
-        if key not in self._queue:
-            self._queue[key] = ZarrChecksumModification(path=key)
-        self._queue[key].paths_to_remove.append(str(path))
+        self._get_path(key).paths_to_remove.append(str(path))
 
     def pop_deepest(self):
```

Comment:
Not strictly needed, but fwiw mypy will not type-check functions that don't have any type annotations for their arguments or return values.
```suggestion
    def pop_deepest(self) -> ZarrChecksumModification:
```